src/compact.rs

```diff
@@ -206,18 +208,39 @@ impl<'de, T> serde::Deserialize<'de> for Compact<T> where T: serde::Deserialize<
 	}
 }
 
+/// Requires the presence of `MaxEncodedLen` when the `max-encoded-len` feature is active.
+// Remove this trait when the feature is removed.
+#[cfg(feature = "max-encoded-len")]
+pub trait MaybeMaxEncodedLen: MaxEncodedLen {}
+#[cfg(feature = "max-encoded-len")]
+impl<T: MaxEncodedLen> MaybeMaxEncodedLen for T {}
+
+/// Requires the presence of `MaxEncodedLen` when the `max-encoded-len` feature is active.
+// Remove this trait when the feature is removed.
+#[cfg(not(feature = "max-encoded-len"))]
+pub trait MaybeMaxEncodedLen {}
+#[cfg(not(feature = "max-encoded-len"))]
+impl<T> MaybeMaxEncodedLen for T {}
+
 /// Trait that tells you if a given type can be encoded/decoded in a compact way.
 pub trait HasCompact: Sized {
 	/// The compact type; this can be
-	type Type: for<'a> EncodeAsRef<'a, Self> + Decode + From<Self> + Into<Self>;
+	type Type: for<'a> EncodeAsRef<'a, Self> + Decode + From<Self> + Into<Self> + MaybeMaxEncodedLen;
 }
 
 impl<'a, T: 'a> EncodeAsRef<'a, T> for Compact<T> where CompactRef<'a, T>: Encode + From<&'a T> {
 	type RefType = CompactRef<'a, T>;
 }
 
+#[cfg(feature = "max-encoded-len")]
+impl<T> MaxEncodedLen for Compact<T> where T: CompactAs, Compact<T::As>: MaxEncodedLen, Compact<T>: Encode {
+	fn max_encoded_len() -> usize {
+		Compact::<T::As>::max_encoded_len()
+	}
+}
+
 impl<T: 'static> HasCompact for T where
-	Compact<T>: for<'a> EncodeAsRef<'a, T> + Decode + From<Self> + Into<Self>
+	Compact<T>: for<'a> EncodeAsRef<'a, T> + Decode + From<Self> + Into<Self> + MaybeMaxEncodedLen
```

Comment:
we can't use conditional compilation on just the constraint here right? That's why we need the MaybeMaxEncodedLen trait

Reply:
I didn't checked directly, but I don't think so.